deployment/postgres-init-db/sql/views/merlin/derived_events.sql

```diff
@@ -0,0 +1,106 @@
+-- Rule 1. An External Event superseded by nothing will be present in the final, derived result.
+-- Rule 2. An External Event partially superseded by a later External Source, but whose start time occurs before the start of said External Source(s), will be present in the final, derived result.
+-- Rule 3. An External Event whose start is superseded by another External Source, even if its end occurs after the end of said External Source, will be replaced by the contents of that External Source (whether they are blank spaces, or other events).
+-- Rule 4. An External Event who shares an ID with an External Event in a later External Source will always be replaced.
+
+create materialized view merlin.derived_events
```

Comment:
I've reworked parts of this view to be simpler.

Summary of changes:
1) Updated formatting to match the Aerie DB Conventions
2) Renamed sub selects to be more clear (willing to discuss)
3) Used a `distinct on` and `order by` clause to replace the `row_number` hack for Rule 4. This allows us to "flatten out" the `t` subselect.
4) Refined `ranges_with_subs` to avoid joining sources in different derivation groups. This also allows the `subsequent_ranges` column to be defined in a simpler way
5) Removed `original_ranges` and `subsequent_ranges` from `ranges_with_subs`. Neither of these columns were selected in the higher queries.

```sql
create materialized view merlin.derived_events_better as
-- "distinct on event_key" and "order by valid_at" satisfies rule 4
-- (only the most recently valid version of an event is included)
select distinct on (event_key)
    output.event_key,
    output.source_key,
    output.derivation_group_name,
    output.event_type_name,
    output.duration,
    output.start_time,
    output.source_range,
    output.valid_at
from (
  -- select the events from the sources and include them as they fit into the ranges determined by sub
  select
    s.key as source_key,
    ee.key as event_key,
    ee.event_type_name,
    ee.duration,
    s.derivation_group_name,
    ee.start_time,
    s.source_range,
    s.valid_at
  from merlin.external_event ee
  join (
    with base_ranges as (
      -- derivation_tb_range orders sources by their valid time
      -- and extracts the multirange that they are stated to be valid over
      select
        external_source.key,
        external_source.derivation_group_name,
        tstzmultirange(tstzrange(external_source.start_time, external_source.end_time)) as range,
        external_source.valid_at
      from merlin.external_source
      order by external_source.valid_at
    ), base_and_sub_ranges as (
      -- ranges_with_subs takes each of the sources above and compiles a list of all the sources that follow it
      -- and their multiranges that they are stated to be valid over
      select
        base.key,
        base.derivation_group_name,
        base.range                                                      as original_range,
        array_remove(array_agg(subsequent.range order by subsequent.valid_at), NULL) as subsequent_ranges,
        base.valid_at
      from base_ranges base
      left join base_ranges subsequent
        on base.derivation_group_name = subsequent.derivation_group_name
        and base.valid_at < subsequent.valid_at
      group by base.key, base.derivation_group_name, base.valid_at, base.range
    )
    -- this final selection (sub) utilizes the first, as well as merlin.subtract_later_ranges,
    -- to produce a sparse multirange that a given source is valid over.
    -- See merlin.subtract_later_ranges for further details on subtracted ranges.
    select
      r.key,
      r.derivation_group_name,
      merlin.subtract_later_ranges(r.original_range, r.subsequent_ranges) as source_range,
      r.valid_at
    from base_and_sub_ranges r
    order by r.derivation_group_name desc, r.valid_at) s
  on s.key = ee.source_key
  and s.derivation_group_name = ee.derivation_group_name
  where s.source_range @> ee.start_time
  order by valid_at desc
) output;
```

Reply:
Following some discussion in slack, this is now in with [update derived_events.sql per Theresa's rewrite](https://github.com/NASA-AMMOS/aerie/commit/a86818f83ddcc5733ec2effa8e4256a96cc85f47).